### ISSUE TITLE
Incompatible function pointer type casts

### ISSUE THREAD

**Issue by vlad902** [mm2/Little-CMS#145]
Some compiler tools and instrumentation like Control Flow Integrity and Emscripten require that function pointers only call functions that have types matching the function pointer type. In enabling Control Flow Integrity for Chrome, I've run across an instance in LCMS that calls some functions with two different function pointer types making it difficult to fix this error. In particular, function pointers are called as both _cmsInterpFn16 (from struct cmsInterpFunction field Lerp16) and _cmsOPTeval16Fn (cast in the call to _cmsPipelineSetOptimizationParameters from OptimizeByResampling). These types [1] differ in the third argument's pointer type. To fix, one of the types could be eliminated and casts applied where necessary to accommodate.

[1]
```
typedef void (* _cmsInterpFn16)(register const cmsUInt16Number Input[],
                                register cmsUInt16Number Output[],
                                register const struct _cms_interp_struc* p);

typedef void     (* _cmsOPTeval16Fn)(register const cmsUInt16Number In[],
                                     register cmsUInt16Number Out[],
                                     register const void* Data);
```

**Comment by mm2** [comment]
Hi, 
From the ISO/IEC 9899:TC3 (c99) spec:

"6.3.2.3 Pointers
1 A pointer to void may be converted to or from a pointer to any incomplete or object
type. A pointer to any incomplete or object type may be converted to a pointer to void
and back again; the result shall compare equal to the original pointer."
 
Maybe I'm wrong on this, but it seems to me that both function pointers should be compatible  because the above requirement. Please let me know if this is not the case.

**Comment by vlad902** [comment]
Later on in 6.3.2.3:
```
8 A pointer to a function of one type may be converted to a pointer to a function of another
type and back again; the result shall compare equal to the original pointer. If a converted
pointer is used to call a function whose type is not compatible with the pointed-to type,
the behavior is undefined.
```

This is checked by UBSan `-fsanitize=function` on a limited set of platforms.

**Comment by mm2** [comment]
Sure, that is the point.The spec says "compatible", and because point 1 function parameters should be compatible. I didn't get any complaint from lints until now.

**Comment by vlad902** [comment]
Convertibility and compatibility are different in the standard. For example, per 6.3.2.3p5/p6 void* is convertible to size_t and back. However, 6.7.5.1p2 states that pointers are compatible if the types they point to are compatible, but void is not compatible with a struct.

This is a common form of UB that generally does what you expect so I'm not surprised that linters have not looked for it. This is an issue for CFI/emscripten because they exploit type information to optimize their respective implementations.

**Comment by mm2** [comment]
Fine, I'm adding this in the todo list for 2.10. 
I released 2.9 one week ago, so it will take some time. About a year, usually.

Thanks for reporting.

**Comment by mm2** [comment]
Now it is fixed.
